// pd-redis-parentize/lib/nomenclt.js
'use strict';
var humanize = require('humanize-plus');
var pluralize = require('pluralize');  //Even though humanize also has pluralize function but it cannot even solve ly to lies..
var phNm = require('./parenthood-nomenclt');

module.exports = {
    bearerFor: function (Child) {
        var capitalizedChildName = humanize.capitalize(Child.modelName());
        return 'bear' + capitalizedChildName;
    },
    counterFor: function (Child) {
        return Child.modelName() + 'Amount';
    },
    sidListFor: function (Child) {
        return Child.modelName() + 'SidsOf';
    },
    childListOf: function (Child) {
        return pluralize(Child.modelName()) + 'Of';
    },
    ifOwns: function (Child) {
        var capitalizedChildName = humanize.capitalize(Child.modelName());
        return 'has' + capitalizedChildName;
    },
    adopt: function (Child) {
        return 'adopt' + humanize.capitalize(Child.modelName());
    },
    abandon: function (Record) {
        return phNm.abandon(Record.modelName());
    }
};
